### ISSUE TITLE
Cannot invoke an object which is possibly 'undefined' even after checking that it isn't undefined

### ISSUE THREAD

**Issue by alfrdmalr** [microsoft/TypeScript#41371]
**TypeScript Version:**  4.0.3

**Search Terms:**
Cannot invoke an object which is possibly 'undefined'
TS2722
possibly undefined callback"

**Code**

```ts
// simple repro interface/function
interface ListArgs {
    items: string[]
    listTransform?: (items: string[], query?: string) => string[];
    filter?: (item: string, query?: string) => boolean;
}

const myFunc = (args: ListArgs) => {
    let items: string[] = args.items;
    const query: string = 'test';

    // the listTransform function is called right after the check without complaint
    if (args.listTransform) {
        items = args.listTransform(items, query);
    }

    // the filter function is handed off after the check, which the compiler doesn't like
    if (args.filter) {
        items = items.filter(item => args.filter(item, query));
    }

    return items;
}

// run the function:
const items: string[] = ['testing', 'this', 'behavior', 'test']
const dumbFilter = (s: string, q?: string): boolean => {
    return s === q;
}
const dumbTransform = (items: string[], q?: string): string[] => {
    console.log("here's the query: ", q);
    return items;
}
console.log(myFunc({items: items, listTransform: dumbTransform, filter: dumbFilter}))
```

**Expected behavior:**
No error is thrown if the object is checked for undefined-ness.
Using a function that _could_ be undefined but has been checked for existence (`args.filter` in this case) should work without explicitly asserting that it isn't undefined. This seems to be the case if the function is called directly after the check (as in `args.listTransform`), but not if it's passed to another function.

**Actual behavior:**
When used as a callback for another function, the check for undefined-ness is seemingly ignored and its type is still "possibly undefined".
What's worked is asserting the function directly i.e. 
```
if (args.filter) {
        // manually assert args.filter isn't null
        items = items.filter(item => args.filter!(item, query));
}
```
 as well as casting by reassigning to a new variable after checking:
```
if (args.filter) {
        // assign type directly
        const f: (item: string, query: string) => boolean = args.filter;  
        items = items.filter(item => f(item, query));
}
```
Both of these approaches work just fine, but I'm wondering why they're necessary. What's different about calling the function right away vs handing it off that causes this error?

**Playground Link:** 
[Playground link for this example](https://www.typescriptlang.org/play?#code/PTAEGcEsFsAcBsCmoBOjYoPahLv+DCjiTSzyAoSAOwBdEUAzAQwGNkAZScWgQRQDm4UAG8KoCaEj1o4AFwRaKGgIDaAXXGT43WgBUUzauEaYU0APwKAFNMSyFPZdTXqANKACOAVwYBPK0VnAQBKUABeAD4glQ0Abi0JRkh4ehRA2xlHJRUPH39ApxUwqNAAI0xMJCMEgF8KClZMY1pQaD8AMW9qVgjQa2ZBeVAuHn4hEuixSVAkVrsHGJcNPsGhADoF8ASZppavXxQ-bOC+gHJ6HjOExKlGfrXwdZ0eAyMTM2gw6ZnJLdWhs9dG9jKZzJl7OA8oc-CEdpJ6rdIPcBoDkqkGN9bjN-uEpDInui0hDoBFoo91kSGCTof4QnDboiZmhaN4UNR8ZC6g09jxOYsist1H1VBdEDwVGcPBcABbcKWgM5lRAy5gAN0gZgVYquml5rQAJt5oGUOik0n1rMNBQI8oUci4QgoKlVEEYyaJbiy2RzhOF-V5ufrQEaTSCPuZLVsTrF3F57cEnUtXB6fhJea7npgBNYAEQyhiIM7CWgFg7+BS5vIM5mIVns-nbCj1DNILM59pdHrWETRxseF76Qygz4KUNlcNg6AeKkoMfG03mhi1ekNECgACiADkACKgABKG4ACvuAPI4ciXq-Xm83tdgcBVby0TXGUCl5itADuZgA1goeWaPloCMbxmHgeA-F4cBwAYWgt0weDvAgy1HgUUY+CGSZPRmOZGxjIUAQ2LZ4XTIDWnyI4CIEc5Lloa4GhxFEKUHSdPixX4-gJIinlY4cI2gEkoXLI4awRRi-mYtElxQDjOIvNpQPAyDQGYGC4NU6SMRQKRwGoC5QGoZD4GxLjIT6LZKRkkkPQpWcAEIaRE2ExIkJlJG9BsSObQD9lYNTaAANUgZh9zdGDIAEahoEQOhUKGdDdHGcBsLTWY63w5MVjxClvN2cjnOo2jxXom4mIeQE+PeKc5M43FNI2KqR3BLZaVE0jQHciRkQqjZZ1q351zUqAovfPxYGQA1IDQVhaEg0yyP2RgbAWai2uOZNsJdaoORyrS0jiCQFsbCyCSs7SbNKRgnMolyOq61A6x9RtuQoIbYAwTA2ALEtVVaBgsDnRoCv8nhgtC8KRui2LaAAdWkGVMGfABhAtWF-eKhESsYsNTW48L7G1soap48skYNbqKvEdVKiTuqkxrgX4mqcPk+qWKZ6rPiE9bXM6unQGDZb+lW5N1uorbKh2njzoO0AkW4vFLNnS7omuhZeZCAXPI5bzEWDRAAA8EEgVhpEg1HEHRgBVagDUQZJqEQA1MeGDDktS-GMsJh0Uz24iCQ6imYSpxU6IYpEGd4znmq+Vm6sVkmgVeZnuda5y+YenrUT6mTQHsgNuntx3nYGnFE+V6yFls-bqQ1jPM+1p6vMD5sgA)
**Related Issues:** 
#31609

**Comment by MartinJohns** [comment]
> **TypeScript Version:** 3.7.x-dev.201xxxxx

You might want to update your TypeScript version. This is already quite old and a preview version.

---

Duplicate of #9998. Would be solved by #11498.

**Comment by alfrdmalr** [comment]
Whoops. managed to miss that part of the issue template. I am running version 4.0.3. 
Either way, thanks for pointing me to those two issues. I think I didn't know quite what I needed to look for and those discussions more thoroughly dive into the weeds on this particular behavior. I'll go ahead and close this.
